File: backend/src/models/cart.model.ts
import { ICart, ICartItemDoc, ICartItemVariationDoc } from '@/types';
import mongoose from 'mongoose';

const CartItemVariationSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    choice: { type: String, required: true, trim: true },
    extraPrice: { type: Number, default: 0 },
  },
  { _id: false }
);

const CartItemSchema = new mongoose.Schema(
  {
    itemId: { type: mongoose.Schema.Types.ObjectId, default: () => new mongoose.Types.ObjectId() },
    productId: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
    quantity: { type: Number, required: true },
    price: { type: Number, required: true },
    note: { type: String, default: null },
    variations: { type: [CartItemVariationSchema], default: [] },
  },
  { _id: false }
);

// --- CARTS ---
const CartSchema = new mongoose.Schema<ICart>(
  {
    storeId: { type: mongoose.Schema.Types.ObjectId, ref: 'Store', required: true },
    cusId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    items: { type: [CartItemSchema], default: [] },
  },
  {
    timestamps: true,
  }
);

// Indexes
CartSchema.index({ cusId: 1, storeId: 1 }, { unique: true });
CartSchema.index({ storeId: 1 });

const CartModel = mongoose.model<ICart>('Cart', CartSchema, 'carts');

// --- CART ITEMS (Separate collection mapping) ---
const CartItemDocSchema = new mongoose.Schema<ICartItemDoc>(
  {
    cartId: { type: mongoose.Schema.Types.ObjectId, ref: 'Cart', required: true },
    productId: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
    quantity: { type: Number, required: true },
    price: { type: Number, required: true },
    note: { type: String, default: null },
  },
  {
    timestamps: true,
  }
);

CartItemDocSchema.index({ cartId: 1 });

export const CartItemModel = mongoose.model<ICartItemDoc>('CartItem', CartItemDocSchema, 'cart_items');

// --- CART ITEM VARIATIONS (Separate collection mapping) ---
const CartItemVariationDocSchema = new mongoose.Schema<ICartItemVariationDoc>(
  {
    cartItemId: { type: mongoose.Schema.Types.ObjectId, ref: 'CartItem', required: true },
    name: { type: String, required: true, trim: true },
    choice: { type: String, required: true, trim: true },
    extraPrice: { type: Number, default: 0 },
  },
  {
    _id: true,
  }
);

CartItemVariationDocSchema.index({ cartItemId: 1 });

export const CartItemVariationModel = mongoose.model<ICartItemVariationDoc>('CartItemVariation', CartItemVariationDocSchema, 'cart_item_variations');

export default CartModel;

File: mobile/lib/features/cart/data/models/cart_model.dart
/// JSON-serializable cart item model matching backend response.
class CartItemModel {
  final String? itemId;
  final String productId;
  final String? productName;
  final String? productImage;
  final double price;
  final int quantity;
  final String? note;
  final List<CartItemVariationModel> variations;

  const CartItemModel({
    this.itemId,
    required this.productId,
    this.productName,
    this.productImage,
    required this.price,
    this.quantity = 1,
    this.note,
    this.variations = const [],
  });

  factory CartItemModel.fromJson(Map<String, dynamic> json) {
    return CartItemModel(
      itemId: json['itemId'] as String?,
      productId: json['productId'] as String? ?? '',
      productName: json['productName'] as String?,
      productImage: json['productImage'] as String?,
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      quantity: json['quantity'] as int? ?? 1,
      note: json['note'] as String?,
      variations:
          (json['variations'] as List<dynamic>?)
              ?.map(
                (e) =>
                    CartItemVariationModel.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
    'itemId': itemId,
    'productId': productId,
    'productName': productName,
    'productImage': productImage,
    'price': price,
    'quantity': quantity,
    'note': note,
    'variations': variations.map((v) => v.toJson()).toList(),
  };
}

/// JSON-serializable cart item variation model.
class CartItemVariationModel {
  final String name;
  final String choice;
  final double extraPrice;

  const CartItemVariationModel({
    required this.name,
    required this.choice,
    this.extraPrice = 0.0,
  });

  factory CartItemVariationModel.fromJson(Map<String, dynamic> json) {
    return CartItemVariationModel(
      name: json['name'] as String? ?? '',
      choice: json['choice'] as String? ?? '',
      extraPrice: (json['extraPrice'] as num?)?.toDouble() ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'choice': choice,
    'extraPrice': extraPrice,
  };
}
